[1단계 - 장기] 하루(구은선) 미션 제출합니다.

eunseongu opened the pull request:
안녕하세요 앤지 😊
7기 하루입니다.
리뷰 잘 부탁드립니다!

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
https://prolog.techcourse.co.kr/studylogs/4120

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

객체지향 생활체조 요구사항을 최대한 지키려고 노력했습니다!
다만 부족한 부분이 있을 것이라 생각해 4점을 선택했습니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

이번 미션에서는 Controller 같은 클래스를 만들지 않고, 기물(Piece)과 보드(Board) 사이의 협력을 통해 기능을 구현하는 것을 목표로 삼았습니다.

이를 위해 `Board`는 기물들의 추상 클래스인 `Piece`가 어디에 위치하는지만 관리하고, 각 기물의 구체적인 종류나 내부 동작 방식은 알지 못하도록 설계했습니다.
또한, 기물의 이동, 이동 검증, 특정 위치에 기물이 존재하는지 여부에 대해서만 책임을 지도록 구성했습니다.

반면, 각 기물의 구현체는 자신만의 이동 방식 검증과 이동 경로 계산에 대한 책임을 갖도록 분리하여,
기물 스스로가 자신의 행동을 정의하고 수행할 수 있도록 구현해 보았습니다.

최대한 `Board`와 `Piece`가 서로의 역할을 침범하지 않으면서 협력하도록 구성했는데,
이러한 설계 방식이 코드에 잘 반영되었는지 피드백 부탁드립니다! 😊

YehyeokBang committed:
docs: 구현할 기능 목록 작성

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 좌표 모델 구현

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 병 모델 구현

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 졸 모델 구현

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 군인 모델 구현

- 졸과 병을 진영으로 구분합니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 마 모델 구현

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 상 모델 구현

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 차 모델 구현

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 포 모델 구현

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 장군, 사 모델 구현

- 현재는 움직임을 구현하지 않습니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 기물 추상화 구현

- 기물은 진영와 음직임 검증을 책임집니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
refactor: 기물 패키지 분리

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 보드판의 기물 두기 기능 구현

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 보드판 초기화 기능 구현

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
refactor: 코드 포맷팅

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 보드판 초기화 및 출력 기능 구현

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 게임 시작 기능 구현

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
refactor: 출력 포맷 변경

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
refactor: 좌표 체계 변경

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 기물의 일반적인 움직임 구현

- 현재 버전에서는 상대 기물을 고려하지 않습니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 다른 진영의 기물을 잡는 기능 구현

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 포 고유 기능 구현

- 포는 포가 아닌 1개의 기물을 넘어야 이동할 수 있습니다.
- 좌표 수평 수직 기준을 변경합니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 차 고유 기능 구현

- 차는 다른 기물을 넘어 이동할 수 없습니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 마 고유 기능 구현

- 마는 다른 기물을 넘어 이동할 수 없습니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 상 고유 기능 구현

- 상은 다른 기물을 넘어 이동할 수 없습니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
feat: 기물 잡는 규칙 추상화

- 같은 진영의 기물을 잡을 수 없습니다.
- 포는 다른 진영이어도 포를 잡을 수 없습니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
refactor: 보드판에서 진영 의존성 제거

- 보드판은 각 기물이 어떤 진영인지 관심이 없습니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
refactor: 포 리팩토링

- 가독성 개선 및 테스트 케이스를 추가합니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
refactor: 차, 상, 마 리팩토링

- 가독성을 개선합니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
refactor: 기존과 같은 곳으로 움직이는 검증 구현

- 보드판에서 같은 곳으로 움직이려는 요청을 검증합니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
refactor: 불필요한 필드 제거

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
refactor: 입출력 기능 리팩토링

- 인덴트 줄이기 및 가독성을 개선합니다.
- 리드미 문서를 보완합니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
refactor: 접근자 메서드 위치 변경

- 단순 getter 위치를 아래로 이동합니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

YehyeokBang committed:
test: 포 테스트 추가

- 포는 포를 넘어서 이동할 수 없습니다.

Co-authored-by: YeonhuLee <yeonnhuu@gmail.com>
Co-authored-by: eunseongu <rndmstjs75@gmail.com>

eunseongu commented on src/main/java/janggi/board/Board.java:
```diff
@@ -0,0 +1,84 @@
+        this.placedPieces = initializeBoard();
+    }
+
+    private Map<Point, Piece> initializeBoard() {
+        Map<Point, Piece> board = new HashMap<>();
+        for (int i = 1; i <= COLUMN; i++) {
+            for (int j = 0; j < ROW; j++) {
+                board.put(new Point(i, j), null);
+            }
+        }
+        return board;
+    }
```
이 부분은 indent를 1로 유지하라는 규칙을 지키지 못했습니다.

다만, indent를 2로 유지했을 때 보드를 열과 행의 사이즈 기준으로 초기화한다는 흐름이 더 직관적으로 보인다고 판단하여,
가독성과 로직의 명확성을 고려해 현재 구조를 선택했습니다.

이 선택이 규칙을 벗어난 것이 맞는 만큼,
보다 나은 방식으로 구조를 개선할 수 있는 방향이 있다면 피드백 부탁드립니다!

soominsohn replied:
네. 저도 행과 열을 초기화하는 코드에 `indent 2`는 허용해도 괜찮다고 생각합니다. 
다만 초기화 할 때 board의 Piece에 null이 올 수 있는데요, null사용은 조심하는게 좋습니다.

eunseongu commented on src/main/java/janggi/piece/Soldier.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Override
+    public void validateMove(Point fromPoint, Point toPoint) {
+        if (isBottom()) {
+            validateJolMove(fromPoint, toPoint);
+            return;
+        }
+        validateByeongMove(fromPoint, toPoint);
+    }
```
장기판에서 ‘초’ 진영의 기물은 아래 방향에 배치되어 있고,
특히 군인 기물은 진영에 따라 이름이 달라지는 특징이 있어 이를 구분해줄 필요가 있었습니다.

그래서 `Camp` 클래스에 `isBottom()` 메서드를 만들어,
‘초’ 진영이 아래 방향임을 명시하고, 이를 기반으로 군인 기물의 이동 방향을 검증하는 방식으로 구현했습니다.

메서드 이름인 `isBottom()`만 보고도 해당 진영의 위치 정보나 이동 방향과 관련된 역할을 유추할 수 있었는지,
혹시 직관적이지 않았다면 이름이나 구조 면에서 어떤 부분을 보완하면 좋을지 피드백 부탁드립니다! 😊

soominsohn replied:
isBottom이라는 이름은 도메인을 어느 정도 아는 사람에게는 충분히 직관적이지만, 
장기 규칙이나 진영 배치를 모르는 사람에게는 `bottom`의 의미가 모호할 수 있어요.
이를 `isJol()`로 변경하는 것도 고려해볼만 하다고 생각이 듭니다.

eunseongu reviewed (commented):
안녕하세요 앤지!

코드와 관련된 질문은 코멘트로 남겨두었습니다!

soominsohn commented on src/main/java/janggi/board/Board.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    private void validatePoint(Point point) {
+        if (point.getX() < 0 || COLUMN <= point.getX() || point.getY() < 0 || ROW <= point.getY()) {
```
Point의 값을 getter로 꺼내와 비교하기보다는, 비교 로직을 Point 객체 내부에 구현해보면 어떨까요?
이렇게 하면 어떤 장점이 있을지 함께 고민해보면 좋을 것 같아요.

eunseongu replied:
Point 내부에서 자신이 특정 범위 안에 있는지를 직접 판단하도록 변경했습니다!

이렇게 변경하면서 외부에서는 Point의 내부 정보를 알 필요 없이
Point 스스로가 자신의 상태를 기반으로 범위 검증을 수행하고 결과를 반환할 수 있게 되었습니다.

이를 통해 외부에서 좌표 정보를 검증하거나 비교하지 않아도 되어 캡슐화가 더 잘 지켜지게 되었다고 생각합니다! 😊

soominsohn commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public void validateCatch(Piece otherPiece) {
+        if (camp == otherPiece.getCamp()) {
```
이 부분도 getter를 사용하지 말고 비교해볼까요?
코드 전체적으로 getter를 사용을 지양하도록 수정해볼까요?

eunseongu replied:
아래와 같이 Piece 클래스에서 otherPiece의 camp 값을 직접 비교하는 방식으로 변경해보았습니다:

```java
public void validateCatch(Piece otherPiece) {
    if (this.camp == otherPiece.camp) {
        throw new IllegalArgumentException("같은 진영의 기물을 잡을 수 없습니다.");
    }
}
```

이 코드를 작성하면서 궁금한 점이 생겼습니다!

Piece 객체에서 otherPiece의 private 필드인 camp에 접근할 수 있는 이유가 같은 클래스이기 때문인데, 
같은 클래스라고 해도 다른 인스턴스의 private 필드에 직접 접근하는 방식이 괜찮은지 궁금합니다.

이런 접근 방식에 대해 주의해야 할 점이 있다면 조언 부탁드립니다!😊

eunseongu replied:
Point에서 사용하던 getter를 제거하기 위해 수직, 수평선 상에 있는 좌표 찾는 로직을 Point 객체에서 스스로 하도록 변경했습니다!

soominsohn replied:
같은 클래스라고 해도, 다른 인스턴스의 상태를 직접 들여다보는 건 어느정도는 캡슐화를 깨는 것 같다고 생각이 들긴 합니다. 
`otherPiece.getCamp()` 처럼 getter를 사용하는 방식이 조금 더 안전할 것 같아요. 😄

soominsohn commented on src/main/java/janggi/piece/Soldier.java:
```diff
@@ -0,0 +1,45 @@
+        if (fromPoint.getY() < toPoint.getY()) {
+            throw new IllegalArgumentException("병은 뒤로 갈 수 없습니다.");
+        }
+        if (Math.abs(toPoint.getY() - fromPoint.getY() + fromPoint.getX() - toPoint.getX()) != 1) {
+            throw new IllegalArgumentException("병은 앞 또는 양 옆으로 한 칸만 움직일 수 있습니다.");
+        }
```
![Pasted Graphic 2](https://github.com/user-attachments/assets/4be3587a-ab09-4312-b4ca-26c3d60362d8)
요 오류가 발생하면 프로그램이 종료되네요. 이 부분도 종료되지 않고 계속 진행되게 바꿔주면 좋을 것 같습니다.

eunseongu replied:
기존에 try-catch 안에서 `playTurn()` 메서드를 실행하면서 예외가 catch 안에서 생기면 재입력이 되지 않았습니다!

이 문제를 해결하기 위해 while문 안에 try-catch를 넣어,
올바른 입력을 받을 때까지 반복되도록 구조를 변경했습니다.

다만 이 경우, try-catch가 while문 안에 들어가면서 indent가 2단계로 깊어지는 문제가 생겼습니다.
그래서 아래와 같이 indent를 1로 줄이는 방식을 고려해봤습니다.
```java
private static void playTurnUntilSuccess(View view, Camp currentTurnCamp, Board board) {
    boolean success = false;

    while (!success) {
        success = tryPlayTurn(view, currentTurnCamp, board);
    }
}

private static boolean tryPlayTurn(View view, Camp currentTurnCamp, Board board) {
    try {
        List<List<Integer>> moveInput = view.readMove(currentTurnCamp);
        playTurn(moveInput, currentTurnCamp, board);
        return true;
    } catch (IllegalArgumentException e) {
        System.out.printf(ERROR_MESSAGE_FORMAT, e.getMessage());
        return false;
    }
}
```

하지만 이 방식은 success라는 불필요한 상태값을 사용하고 있어 좋은 방법 같지 않다는 생각을 했습니다!
그래서 우선 indent를 2로 유지해두었는데, 더 좋은 방법이 있을지 조언해주시면 감사하겠습니다 😊

soominsohn replied:
```
    private static void playGame(View view) {
        Board board = BoardGenerator.generate();
        Camp currentTurnCamp = FIRST_TURN_CAMP;
        while (true) {
            view.displayBoard(board.getPlacedPieces());
            currentTurnCamp = playTurnUntilSuccess(view, currentTurnCamp, board);
        }
    }

    private static Camp playTurnUntilSuccess(View view, Camp currentTurnCamp, Board board) {
        try {
            List<List<Integer>> moveInput = view.readMove(currentTurnCamp);
            playTurn(moveInput, currentTurnCamp, board);
            return currentTurnCamp.reverse();
        } catch (IllegalArgumentException e) {
            System.out.printf(ERROR_MESSAGE_FORMAT, e.getMessage());
            return currentTurnCamp;
        }
    }
```
하루의 Application 코드에 있는 코드를 조금 긁어왔는데, 메서드명은 바꿔야할 것 같지만 이렇게 바꿔보면 어떨까요?
성공하면 턴을 바꿔서 리턴하고, 실패하면 턴을 바꾸지 않게 리턴해서 계속 진행할 수 있게요!

soominsohn commented on src/main/java/janggi/view/View.java:
```diff
@@ -0,0 +1,87 @@
+    public void displayStartBanner() {
+        System.out.println("""
+                
+                ====================================
+                    Welcome to the Janggi Game!
+                ====================================
+                 {가로축 번호}{세로축 번호} 좌표를 사용합니다.
+                    예) 41은 왼쪽에서 4번째 아래에서
+                        첫번째 기물을 의미합니다.
+                      게임은 초나라부터 시작됩니다.
+                ====================================
+                """);
```
![image](https://github.com/user-attachments/assets/f7d8acc0-a6a2-44fb-9dc7-09cc2c358020)
옛날 콘솔게임 같이 귀엽네요 😄

soominsohn commented on src/test/java/janggi/board/BoardTest.java:
```diff
@@ -0,0 +1,155 @@
+        // when & then
+        assertThatCode(() -> board.placePiece(point, piece))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("기물의 위치는 9 x 10 영역을 벗어날 수 없습니다.");
```
hasMessage로 예외 메세지를 직접 비교하게 되면 어떤 단점이 생길 수 있을까요?
수정할 필요는 없고 궁금해서 물어보는겁니다!

eunseongu replied:
예외 메시지가 변경되는 경우 도메인 코드와 테스트 코드의 예외 메시지를 모두 직접 변경해주어야 한다는 단점이 생길 수 있습니다!

이를 개선하기 위해 예외 메시지를 상수로 모아서 관리하는 방법을 생각해봤습니다.
다만 이 방식은 테스트 코드에서 어떤 메시지를 검증하는지 직관적으로 파악하기 어렵다는 단점이 있었습니다.

이 부분에 대한 앤지의 의견은 어떠신가요?

soominsohn replied:
네 맞아요! 예외 메시지는 자주 바뀔 가능성이 있기 때문에 변경될 때마다 일을 두 번 해야할 수 있어요!
단점을 알고가시면 좋을 것 같아 언급했습니다. 😄 
확인해주셔서 고마워요!

soominsohn commented on src/test/java/janggi/piece/ElephantTest.java:
```diff
@@ -0,0 +1,114 @@
+        Elephant elephant = new Elephant(camp, board);
+        Point fromPoint = new Point(5, 5);
+        Point toPoint = new Point(toX, toY);
+        board.placePiece(new Point(5, 6), new Soldier(Camp.CHU, board));
+        board.placePiece(new Point(6, 5), new Soldier(Camp.CHU, board));
+        board.placePiece(new Point(5, 4), new Soldier(Camp.CHU, board));
+        board.placePiece(new Point(4, 5), new Soldier(Camp.CHU, board));
```
`test fixture`에 대해 학습해보고 적용해보면 좋을 것 같아요.

eunseongu replied:
기물에서 보드에 대한 의존성을 제거하면서
기존에는 `board.placePiece()`를 통해서만 테스트가 가능했던 문제도 함께 개선할 수 있었습니다!

테스트를 작성하면서 기물의 상태나 위치 설정을 반복적으로 처리해야 했는데
test fixture를 적용했다면 중복 코드를 줄이고 효율적으로 테스트를 구성할 수 있었을 것 같습니다.

다음에 비슷한 상황이 생기면 test fixture를 적용해서 더 좋은 테스트를 작성해보겠습니다 👍

soominsohn replied:
네 좋아요! test fixture를 사용하면 가독성도 좋아지고 작성해야할 코드양도 줄어들어 귀찮음이 줄어든답니다.. 👍

soominsohn commented on src/main/java/janggi/piece/Elephant.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    private boolean isNextPointOnHorizontal(Point fromPoint, Point toPoint) {
+        return fromPoint.calculateXDistance(toPoint) == 3;
```
숫자 `3`이 의미하는게 뭘까요? 매직넘버 사용은 지양하고 의미있는 상수로 관리해보면 어떨까요?

eunseongu replied:
넵!
‘상’ 기물이 3만큼의 차이를 이동할 수 있는 이유는
직선으로 한 칸 + 대각선으로 두 칸을 이동해야 하기 때문입니다.

이를 반영하여 기물의 이동을 상수로 표현하여 변경했습니다!

```java
 private static final int STRAIGHT_STEP = 1;
 private static final int DIAGONAL_STEP = 2;
 private static final int ELEPHANT_MOVE_DISTANCE = STRAIGHT_STEP + DIAGONAL_STEP;
```

soominsohn commented on src/main/java/janggi/piece/Soldier.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Override
+    public PieceSymbol getPieceSymbol() {
```
PieceSymbol의 패키지 위치는 어떻게 될까요?
미쳐 도메인으로 옮기지 못한 객체일까요?
아니면 도메인 객체가 아닌 다른 객체일까요?

eunseongu replied:
구현 당시에 PieceSymbol 클래스를 어떤 패키지에 넣을지 결정하지 못했습니다..!

Piece와 관련된 정보를 담고 있으니 Piece 패키지로 옮겼습니다!

eunseongu replied:
구현 당시에 PieceSymbol 클래스를 어떤 패키지에 두는 게 적절할지 고민했지만 결정하지 못했습니다 😅

Piece의 정보를 담고 있으니 piece 패키지로 옮기려고 하였으나,
PieceSymbol은 Cannon에서 같은 포인지를 비교하는 로직 외에는 View와 연관된 부분이 많다고 판단했습니다.
그래서 우선 View 패키지로 옮겨둔 상태입니다!
```java
 @Override
    public void validateCatch(Piece otherPiece) {
        super.validateCatch(otherPiece);
        if (getPieceSymbol() == otherPiece.getPieceSymbol()) {
            throw new IllegalArgumentException("포는 포를 잡을 수 없습니다.");
        }
    }
```

같은 포인지를 비교할 때, 뷰를 위해 만든 클래스인 PieceSymbol을 사용하는 대신 더 좋은 방식으로 개선할 수 있지 않을까 생각합니다.

그래서 뷰 로직과의 분리를 위해 새로운 Enum을 만드는 것에 대해 생각해봤는데, 
현재 생각한 방식에서는 아래와 같이 단순한 값의 나열이고 그렇다면 PieceSymbol과 중복되는 부분이 많아 보입니다.
```java
public enum PieceCategory {

    CANNON,
    CHARIOT,
    ELEPHANT,
    GENERAL,
    GUARD,
    HORSE,
    SOLDIER;
}
```

혹시 이 부분에 대해 더 나은 구조나 방법이 있을지 조언해주시면 감사하겠습니다!

soominsohn replied:
domain 패키지에 있는 Piece들에서 view패키지에 있는 PieceSymbol를 사용하면 domain에서 view의 객체를 의존하게 됩니다.
도메인 로직에서는 도메인에서 사용하는 PieceCategory 사용하고 PieceSymbol은 출력 책임만 가지도록 View에 배치하는게 자연스러워 보입니다.
enum이 하는 행위가 비슷하여 중복이라 생각할 수 있지만 domain, view의 책임이 무엇이고 domain에서 왜 view를 의존하면 안되는지에 대해 고민해보면 좋겠습니다.

soominsohn commented on src/main/java/janggi/Point.java:
```diff
@@ -0,0 +1,65 @@
+package janggi;
```
Point 패키지 위치 확인 부탁드립니다 😄

eunseongu replied:
넵! 
Board가 좌표를 관리하고 있기 때문에 board 패키지로 옮겼습니다!

soominsohn commented on src/main/java/janggi/Point.java:
```diff
@@ -0,0 +1,65 @@
+        this.y = y;
+    }
+
+    public Point(String value) {
+        if (value.length() != 2) {
+            throw new IllegalArgumentException("잘못된 좌표 입력입니다.");
+        }
+        try {
+            String[] split = value.split("", -1);
+            this.x = Integer.parseInt(split[0]);
+            this.y = Integer.parseInt(split[1]);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("숫자만 입력 가능합니다.");
+        }
+    }
```
`xy,x'y'` 형식의 입력은 View의 요구사항에 가까워 보여요. 쉼표를 split하는 처리는 도메인 객체보다는 View 쪽에서 처리하는게 좋을 것 같습니다.

eunseongu replied:
넵!
두 경우 모두 잘못된 입력에 대한 처리이기 때문에 Point에서 처리하는 것보다 View에서 입력을 받자마자 처리하는 것이 더 적절하다고 생각합니다!

다만 검증을 위해 파싱하는 과정도 View에 포함되었기 때문에 파싱된 값을 전달해주고자 `List<List<Integer>>`를 사용하게 되었는데, 어떤 위치에 어떤 값이 있는지를 표현하기 어렵다는 단점이 있었습니다.

View는 최대한 사용자의 입력에 대한 처리만 할 수 있도록 하고자 현재 방식을 택했는데, View에서 Point 객체를 만들어서 생성해주는 방식이 더 적절할까요?

이런 경우 앤지는 어떤 방식을 사용하시는지 궁금합니다!

soominsohn replied:
이 부분은 제가 착각하고 리뷰를 드렸네요!
`xy` 이렇게 Point 하나를 입력으로 받는 생성자인데 `xy,x'y'`처럼 Point 두개가 들어온다고 착각을하고 리뷰를 드렸습니다;;
다양한 입력에 대해 객체를 생성하는 것도 생성자의 역할인데요, 다시 확인해보니 이전에 작성해주신 코드도 괜찮다고 생각이 듭니다. 
또 View에서는 입력에 대한 기본적인 유효성 검사만(값을 아예 입력하지 않고 엔터를 치는 경우 등)을 하고, Point객체는 지금처럼 Application에서 만들어주면 좋을 것 같습니다! 혼란스럽게 해서 미안해요! 😢

soominsohn commented on src/main/java/janggi/board/Board.java:
```diff
@@ -0,0 +1,84 @@
+            throw new IllegalArgumentException("해당 위치에서 기물을 찾을 수 없습니다.");
+        }
+        return piece;
+    }
```
Piece에 null이 올 수 있군요. 
null을 직접 다룰 때 어떤 문제점이 올 수 있을까요? 
또 어떻게 문제점을 개선해 볼 수 있을까요?

eunseongu replied:
null을 참조하는 Piece에 접근하게 되는 경우 NullPointException이 발생하게 됩니다.
NullPointException은 런타임 상황에서 발생하기 때문에 미리 예방하지 못하고 프로그램이 작동을 멈춰야 오류를 알 수 있다는 문제가 있습니다.
또한 null을 사용하면 현재 코드와 같이 null인지 검증하고, 처리해주는 로직이 필요합니다.


이를 해결하기 위해 두 가지 방식을 생각했는데, 어떤 방식이 더 좋을지 고민이 되어 모두 시도하고 장단점을 정리해보았습니다.

---
### 1
첫 번째로는 비어있는 기물을 표현하기 위해 Piece를 상속받은 빈 기물 객체를 만들어보았습니다.

이 방식의 장점은 "비어 있음"이라는 상태를 객체로 표현할 수 있다는 점입니다.
기물이 존재하지 않는 경우에도 해당 칸을 객체로 다룰 수 있어서 `piece.isEmpty()`럼 와 같이 **객체에 메시지를 보내서 상태를 판단하니 null인지 아닌지 확인하는 조건문을 줄일 수 있었습니다.
이를 통해 각 기물들이 자신이 비었는지, 이동 가능한지 등을 스스로 판단하게 되었습니다.

하지만 Empty 객체를 표현하기 위해 `Camp.EMPTY`, `PieceSymbol.EMPTY` 같은 실제로는 의미가 없는 값들을 추가로 정의해야 했습니다.
또한 `Camp.reverse()`에서 자기 자신을 반환하거나 `validateMove()`나 `validateCatch()`는 아무 일도 하지 않는 등의 의미 없는 코드가 생겼습니다.

불필요한 행동이 늘어난다는 단점이 크게 느껴져 두 번째 방식을 시도했습니다.

---
### 2
현재 보드에서 map을 사용하고 있기 때문에 초기화 시에 null을 넣지 않고 해당 좌표에 기물이 있는지 확인할 때만 `containsKey()`를 사용하도록 했습니다.

하지만 여전히 해당 위치에 접근하기 전에 존재 여부를 검사해야 한다는 번거로움이 있습니다.
또한, 기물이 비어 있는지 확인할 때 1번 방식과 달리 `isEmpty()`처럼 객체에 메시지를 던지는 게 아니라서, 의도가 덜 드러난다고 생각했습니다.

그럼에도, 빈 기물 객체를 만드는 방식과 달리 비어있는 좌표는 아예 저장하지 않기 때문에 불필요한 데이터를 저장하지 않습니다.
이로 인해 사용하지 않는 enum이나 메서드가 생기지 않는다는 장점이 느껴져서 현재 코드에 반영해보았습니다!


혹시 더 좋은 구조나 피드백 해주실 내용이 있다면 부탁드립니다!

soominsohn replied:
네! 저도 하루가 반영해주신 2번 방법이 괜찮다고 생각이 듭니다.
만일 빈 기물 객체를 꼭 나타내야 한다면 1번 방법이 괜찮겠지만, 지금은 굳이 필요 없어 보여요!
보드의 자료형으로 Map을 사용하고 있으니 Map이 제공해주는 containsKey()를 사용하는 것도 좋아보입니다!
null은 늘 문제를 야기할 수 있기 때문에 null 비교를 지양하고 이를 해결할 수 있는 다른 방법이 있는지 체크해보면 좋아요! 😄 
생각 공유해주셔서 고맙습니다!

soominsohn commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -0,0 +1,45 @@
+public abstract class Piece {
+
+    private final Camp camp;
+    private final Board board;
```
> 어떤 부분에 집중하여 리뷰해야 할까요?
> 이번 미션에서는 Controller 같은 클래스를 만들지 않고, 기물(Piece)과 보드(Board) 사이의 협력을 통해 기능을 구현하는 것을 목표로 삼았습니다.
> 
> 이를 위해 Board는 기물들의 추상 클래스인 Piece가 어디에 위치하는지만 관리하고, 각 기물의 구체적인 종류나 내부 동작 방식은 알지 못하도록 설계했습니다.
> 또한, 기물의 이동, 이동 검증, 특정 위치에 기물이 존재하는지 여부에 대해서만 책임을 지도록 구성했습니다.
> 
> 반면, 각 기물의 구현체는 자신만의 이동 방식 검증과 이동 경로 계산에 대한 책임을 갖도록 분리하여,
> 기물 스스로가 자신의 행동을 정의하고 수행할 수 있도록 구현해 보았습니다.
> 
> 최대한 Board와 Piece가 서로의 역할을 침범하지 않으면서 협력하도록 구성했는데,
> 이러한 설계 방식이 코드에 잘 반영되었는지 피드백 부탁드립니다! 😊

네. 보드는 기물의 위치와 기물들의 이동검증에 대한 책임을지고 기물은 자신만의 이동 검증에 대한 책임을 지게 분리한 것은 좋습니다. 
하지만 보드에서도 기물을 필드로 가지고 있고, 기물에서도 보드를 필드로 가지고 있어요. 이렇게 되면 어떤 문제점이 생길 것 같나요?

eunseongu replied:
보드와 기물이 서로를 참조하게 되면서 순환 참조가 발생합니다.

순환 참조를 가지게 되면서 두 객체 간의 의존성이 복잡해지고, 결합도가 높아집니다.
또한 객체 생성 시점에서 무한루프에 빠지게 될 수도 있습니다.

실제로 보드를 초기화할 때 `null`을 사용하지 않기 위해 `Piece`를 상속받은 빈 기물 객체를 만들었는데,
보드를 초기화할 때 기물이 필요하고, 기물을 생성할 땐 보드가 필요한 상황이 생겼습니다.

이를 개선하기 위해 순환 참조 구조를 제거하고자 했습니다.
보드는 기물의 좌표를 관리해야 하므로, 기물에 대한 의존성은 필요하다고 판단했습니다.

반면 기물은 보드의 정보를 몰라도 자신의 이동 경로나 검증 로직만 처리하면 충분하다고 생각하여 기물에서 보드에 대한 의존성을 제거했습니다.

혹시 이 구조에 대해 더 나은 접근법이나 보완할 점이 있을지 조언해주시면 감사하겠습니다!😊

soominsohn replied:
수정 잘해주신 것 같아요!
설계에 대한 고민도 깊게 해주셨고 순환참조에 대한 생각 정리도 잘 하신 것 같아요! 
공유해주셔서 고맙습니다!

soominsohn reviewed (changes requested):
안녕하세요 하루!
장기 미션을 같이 진행하게 될 앤지라고 합니다.
잘 부탁드려요!
코멘트 남겨두었으니 확인 부탁드립니다. 😄

eunseongu committed:
refactor: 패키지 분리

eunseongu committed:
refactor: 좌표 입력 값 처리 방식 변경

eunseongu committed:
refactor: 패키지 분리

eunseongu committed:
refactor: 기물에서 보드판 의존성 제거

eunseongu committed:
refactor: 재입력 로직 수정

- 예외가 2번 연속으로 생기면 재입력이 되지 않는 로직을 수정합니다.

eunseongu committed:
refactor: 좌표값 범위 검증 로직을 좌표가 스스로 하도록 변경

eunseongu committed:
refactor: 기물 객체 리팩토링

- 기물을 잡을 수 있는 지 검증하는 로직에서 getter를 사용하지 않고 진영을 비교하도록 변경합니다.
- 군인 기물의 졸, 병을 구분하기 위한 메서드 이름을 변경합니다.

eunseongu committed:
refactor: 보드에서 null값을 사용하지 않도록 변경

eunseongu committed:
refactor: 기물의 이동 거리 상수화

eunseongu committed:
refactor: 수직, 수평선 상에 있는 좌표를 좌표 객체가 스스로 찾도록 변경

eunseongu committed:
refactor: 메서드명, 변수명 변경

eunseongu committed:
refactor: 테스트 변수명, 컨벤션 변경

eunseongu committed:
test: 경로 찾는 테스트 추가

soominsohn commented on src/main/java/janggi/Application.java:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    private static void playGame(View view) {
+        Board board = BoardGenerator.generate();
+        Camp currentTurnCamp = FIRST_TURN_CAMP;
+        while (true) {
+            view.displayBoard(board.getPlacedPieces());
+            playTurnUntilSuccess(view, currentTurnCamp, board);
+            currentTurnCamp = currentTurnCamp.reverse();
+        }
+    }
```
@eunseongu 
토끼가 조언해준 요 부분은 꼭 반영해주시면 좋을 것 같아요!
무한루프의 위험에 대해서도 공부해보면 좋아요 😄

soominsohn commented on src/main/java/janggi/board/Board.java:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+
+    public void move(Point from, Point to) {
+        validateMoveRequest(from, to);
+        Piece movingPiece = peek(from);
+        validateMovable(movingPiece, from, to);
+        validateRoute(movingPiece, from, to);
+        validateCatchable(movingPiece, to);
+        executeMove(movingPiece, from, to);
+    }
```
1. 사용자가 입력한 from위치에 기물이 있는지 찾고
2. 해당 기물이 to 위치로 이동할 수 있는지 검증하고
3. 루트에 있는 기물들을 지나갈 수 있는지 검증하고
4. 목적지에 있는 기물을 잡을 수 있는지 검증하고
5. 모두 가능하다면 움직인다!   
너무 좋은 흐름인 것 같아요! 파악하기도 수월했습니다 👍

soominsohn reviewed (approved):
안녕하세요 하루~!
1단계 리팩토링 정말 잘 해주셨네요!
의도한 동작을 코드로 잘 작성해주신 것 같아요.
반영해주셨으면 하는 부분들을 코멘트로 남겨놨으니 2단계 하시면서 같이 반영 부탁드려요~!
